src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java

```diff
@@ -20,6 +21,47 @@ public ConnectedComponentsCalculator(Tangle tangle, int maxTransactions) {
         this.random = new SecureRandom();
     }
 
+    public Collection<Set<Hash>> getConnectedComponents(Set<Hash> transactions) throws Exception {
+        Set<Hash> unvisited = new HashSet<>(transactions);
+        List<Set<Hash>> result = new ArrayList<>();
+
+        // Outer loop iterates once for each component
+        while (!unvisited.isEmpty()) {
+            Set<Hash> currentComponent = new HashSet<>();
+
+            // Perform DFS scan to find all elements in this component
+            Deque<Hash> stack = new ArrayDeque<>();
+            stack.push(unvisited.iterator().next());
+            while (!stack.isEmpty()) {
+                Hash currentHash = stack.pop();
+
+                if (!currentComponent.contains(currentHash)) {
+                    unvisited.remove(currentHash);
+                    currentComponent.add(currentHash);
+
+                    getNeighbors(currentHash).stream()
+                        .filter(t -> unvisited.contains(t))
+                        .forEach(stack::push);
+                }
+            }
+
+            result.add(currentComponent);
+        }
+
+        return result;
+    }
+
+    private Collection<Hash> getNeighbors(Hash hash) throws Exception {
```

Comment:
In graph theory these vertices are called _adjacent_, so maybe `getAdjacentTransactions` would be preferable especially as neighbors is already used in the IRI context.